File: codes/Aiamjay/Week4-Day4/55. Jump Game.py
# encoding = utf-8
import numpy as np


class Solution:
    def solution(self, nums):
        length = len(nums)
        if length < 2:
            return True
        dp = [0] * length
        dp[0] = nums[0]
        for i in range(1, length):
            flag = False
            for j in range(i - 1, -1, -1):
                if i <= dp[j]:
                    dp[i] = i + nums[i]
                    if dp[i] >= length - 1:
                        return True
                    flag = True
                    break
            if not flag:
                return False
        return True

    def test_solution(self):
        nums = [0]
        print(self.solution(nums))

        nums = [1]
        print(self.solution(nums))

        nums = [2,3,1,1,4]
        print(self.solution(nums))

        nums = [3,2,1,0,4]
        print(self.solution(nums))

        nums = [1, 1, 1, 1, 1, 1, 1]
        print(self.solution(nums))

    def random_test(self):
        candidates = range(5)
        nums = np.random.choice(candidates, size = 12, replace=True)
        print('[', ','.join([str(item) for item in nums]), ']')
        print(self.solution(nums))

if __name__ == "__main__":
    s = Solution()
    s.test_solution()
    s.random_test()

File: codes/Shuofei/week 4/55_Jump Game.java
//dp, O(n^2)/O(n)
class Solution {
    public boolean canJump(int[] nums) {
        if (nums.length == 0) return false;
        boolean[] f = new boolean[nums.length];
        f[0] = true;
        for (int i = 0; i < nums.length; i++) {
            if (f[i]) {
                for (int j = 1; j <= nums[i]; j++) {
                    if ((i + j) >= nums.length)
                        break;
                    f[i + j] = true;
                }
            }
        }
        return f[nums.length - 1];
    }
}

//dp, O(n)/O(1)
class Solution {
    public boolean canJump(int[] nums) {
        if (nums.length == 0) return false;
        int max = 0;
        for (int i = 0; i < nums.length; i++) {
            if (i > max) 
                return false;
            max = Math.max(i + nums[i], max);
            if (max >= nums.length - 1)
                return true;            
        }
        return true;
    }
}